### PR TITLE
Improve big io load

### PR DESCRIPTION
**Pull request by VasilMalinov**
the pull request includes the following JIRA commits:
http://jira/browse/SSV-17628 - BSOD ZFSin!spl_get_or_create_hostid+0xdb

http://jira/browse/SSV-17666 - BSOD ZFSin!ZFSin!CreateReparsePoint+13b

Root cause analysis of BSOD-ZFSin!wzvol_WkRtn+223 and implementation
http://jira/browse/SSV-17713
http://jira/browse/SSV-17677

Root cause analysis of BSOD-ZFSin!dispatcher+0x38
http://jira/browse/SSV-17778

### CODE REVIEW

**Comment by datacore-tlaurent** [comment]
Vasil, I looked at the changes and can't see the gain in performance with this one:

<<
The function CreateReparsePoint was changed. The type of a parameters…
… SubstituteName and PrintName were changed from LPCWSTR to UNICODE_STRING. All internal calculations was replaced with structure field Length and use buffer directly.
>>

If we do not gain anything we should not change that code.
Can you explain if I missed something about this particular part of your change?.

**Comment by VasilMalinov** [comment]
Hi Thierry,

                This issue appear sporadically, when I try to create new ZPool.
                Yes, this change does not improve performance, but I don’t what to create new branch only for this commit.
                From other side, all commits affect stabilization of the ZFS, and they helps to see the bottle necks in a ZFS.
                I take a proposal from you, how I can rename the branch.
                From your demo, I understand that changes above does not affect over your idea for the ZFS changes.
                If you want, we can propose these changes to Jorgen that will be good for our activity in the ZFS community.

Best Regards:
Vasil Malinov

**Comment by datacore-tlaurent** [comment]
I just need to understand what was the issue with the routine CreateReparsePoint().  You say that it happens sporadically but there is no mention of what the issue was.  Can you elaborate a little on what the issue was in this routine?  
This way we can also add it to the commit comment when you will propose those changes to Jorgen.

**Comment by VasilMalinov** [comment]
Hi Thierry,

The issue in the routine was the not null terminate string. Sometimes the buffer was not zeroed and as result we have access to out of boundary of this buffer (exam. SubstituteNameLength = (USHORT)wcslen(SubstituteName) * sizeof (WCHAR) was with wrong size. The calculation was more that 1024). In conclusion we have unpredicted algorithm in the future.


Best Regards:
Vasil Malinov

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c [hunk 1]
```diff
@@ -4774,6 +4775,23 @@ dispatcher(
...
+	zmo = DeviceObject->DeviceExtension;
+
+	extern PDRIVER_DISPATCH STOR_MajorFunction[IRP_MJ_MAXIMUM_FUNCTION + 1];
+
+	if (DeviceObject != ioctlDeviceObject) {
+		if (zmo == NULL || ((zmo->type != MOUNT_TYPE_DCB) && (zmo->type != MOUNT_TYPE_VCB))) {
+			if (STOR_MajorFunction[IrpSp->MajorFunction] != NULL) {
+				//dprintf("Relaying IRP to STORport\n");
+				return STOR_MajorFunction[IrpSp->MajorFunction](DeviceObject, Irp);
+			}
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: if NULL then no reason to continue below in the routine. we should jump at the end of the routine and move this code here:
// Got a request we don't care about?
			Status = STATUS_INVALID_DEVICE_REQUEST;				Status = STATUS_INVALID_DEVICE_REQUEST;
			Irp->IoStatus.Information = 0;				Irp->IoStatus.Information = 0;

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c [hunk 2]
```diff
@@ -4782,39 +4800,25 @@ dispatcher(
...
-		mount_t *zmo = DeviceObject->DeviceExtension;
 		if (zmo && zmo->type == MOUNT_TYPE_DCB)
 			Status = diskDispatcher(DeviceObject, Irp, IrpSp);
 		else if (zmo && zmo->type == MOUNT_TYPE_VCB)
 			Status = fsDispatcher(DeviceObject, Irp, IrpSp);
 		else {
 
-			extern PDRIVER_UNLOAD STOR_DriverUnload;
-			extern PDRIVER_DISPATCH STOR_MajorFunction[IRP_MJ_MAXIMUM_FUNCTION + 1];
 			if (STOR_MajorFunction[IrpSp->MajorFunction] != NULL) {
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: this has been checked above. no reason to recheck here.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c [hunk 3]
```diff
@@ -750,8 +750,8 @@ VOID InitVpb(__in PVPB Vpb, __in PDEVICE_OBJECT VolumeDevice)
 										+0xffff928400000000
 
 */
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c]: add a comment to justify the change, like: the LPCWSTR could be non-NULL terminated when passed and causes the buffer to be overran.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c (hunk 1 above)
Comment (VasilMalinov) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: OK. I will do that.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c (hunk 2 above)
Comment (VasilMalinov) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: This code I will move above. Where you proposed.

File: ZFSin/zfs/include/sys/zvol.h [hunk 4]
```diff
@@ -76,6 +76,9 @@ typedef struct zvol_state {
 	/* 'rdiskX' name, use [1] for diskX */
 	uint8_t zv_target_id;
 	uint8_t zv_lun_id;
+	/* Variables involved to LUN reset mechanism. */
+	uint32_t zv_requests_processing_count; /* How meny SRBs are processing*/
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/include/sys/zvol.h]: zv_requests_processing_count is truly not used for something significant. just inc and dec without any special processing if 0 or not so we should remove that.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c (hunk 3 above)
Comment (VasilMalinov) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c]: I will change UNICODE_STRING to PCUNICODE_STRING, and will add your justification propose, why the  LPCWSTR was changed.

File: ZFSin/zfs/module/zfs/zfs_windows_zvol_scsi.c [hunk 5]
```diff
@@ -792,28 +808,42 @@ wzvol_WkRtn(__in PVOID pWkParms)                          // Parm list pointer.
 		goto Done;
 	}
 
-
 	if (sectorOffset >= zv->zv_volsize) {      // Starting sector beyond the bounds?
 		dprintf("%s: invalid starting sector: %d\n", __func__, startingSector);
 		status = SRB_STATUS_INVALID_REQUEST;
 		goto Done;
 	}
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_windows_zvol_scsi.c]: Global comment about the entire processing for the reset handling: 
the way you have done it runs the risk of never (or delaying for a long time) clearing zv_request_cancellation if you receive a lot of data because each I/O is queued in a work item.  zv_requests_processing_count could stay above 0 for a long time.  Normally the initiator stops sending I/O when triggering a RESET but remember that the RESET will be replied right away with a completed status so the initiator could flood the LUN with I/Os after that and this could prevent your count from going to 0 , which in practice could leave you in the cancelling state virtually "forever" or too long at least.  also, after the initiator gets back form the RESEt command it sent it expects further I/Os to *not* be cancelled but in your scheme those will be and that is not correct. 
Ideally ZFS should not use work items to queue the I/Os or they should also be queued in a control queue so the RESET processing would mark them all as cancelled and the srb status would be set later accordingly.  That would be too much change right now.  
At least, if you notice I/Os should be cancelled you should not do the processing *after* you send the read/write I/O but earlier in this routine.  There's no point sending the I/O in zvol_read_win or zvol_write_win if you have the cancel bit set.   I would add that if the cancel bit is set in the zv we should not even queue the I/O in the work item at all and fail it right away with SRB status BUSY then.  No need to queue them.
And then lastly the way you coded the wzvol_reset_notification will probably not work if you get a SRB_FUNCTION_RESET_DEVICE because in that case the LUN value in the Srb is not set or should ignored.  All LUNs in the target should set the zv_request_cancellation bit.  In the current implementation of ZFS there is only lun 0 for each target so it will work somewhat but the search should then specify 0 in that case, not the Lun number in the Srb for a reset device.  The logic will also break if one day someone really uses the lun numbers under the target so I would right away change that wzvol_reset_notification routine for RESET DEVICE to locate all zv's under the same target id (no matter the Lun number) and set the cancel bit for all of those.  It's going to be one today but the logic is future-proof.
In summary after the above I would say that look at the condition that will make you exit the zv_request_cancellation state to be sure you have a deterministic way of getting out of it.  Only I/Os received before the RESET command should be cancelled, not the ones after.

**Comment by VasilMalinov** [comment]
Hi Thierry,

                Thank you for your remarks and suggestions. You are right for zv_request_cancellation flag.
                I will add change that you recommend me. I have in beginning the similar idea, but after testing, I left minimal changes.
The reason of that is:
                The algorithm which works under the  zvol_read_win or zvol_write_win functions.
                The SRBs which in the WorkerThread queue go to ALV tree if they memory range is free, otherwise they begin wait someone to free requested memory range.
                I want to notice that you a very right about  to many changes at the moment. For that I create Jira in which I described the situation.

                For your lastly remark, I need to think for accurately design, because now I use current functionality of the ZFS.


Best Regards:
Vasil Malinov

**Comment by datacore-tlaurent** [comment]
Modification of the RESET logic has been reworked and sent as a separate PR #211
All other modifications in this PR have already been sent and merged in ZFS master.
Closing.
